### PR TITLE
#23440 fix Refund for bundle product without receiving product back e…

### LINKED ISSUE #23440: Refund for bundle product without receiving product back
### Preconditions (*)

1. 2.3.1
2. php7.2

### Steps to reproduce (*)

1. Create bundle product with at least 2 products
2. Create order with bundle product (eg. 50€)
3. Invoice Order
4. Shop Order
5. Create Credit Memo without taking the product back in stock and only refund amount (eg. 10€)
Unbenannt

Unbenannt2

### Expected result (*)

1. Credit Memo for amount 10€ created

### Actual result (*)

1. Error: "The most money available to refund is €xx.xx.
Eventhou the refund was below the mentioned amount!!!

Unbenannt3

### PR DESCRIPTION
**Pull request by KiuNguyen**
### Description (*)

Fix Refund for bundle product without receiving the product back getting error

### Fixed Issues (if relevant)

1. Fixes magento/magento2#23440: Refund for bundle product without receiving product back

### Manual testing scenarios (*)

1. Create a bundle product with at least 2 products
2. Create order with bundle product
3. Invoice Order
4. Shop Order
5. Create Credit Memo without taking the product back in stock and only refund amount:
- Change all qty to 0
- Add adjustment refund: 10$
- Click on button update qty

Screenshot - 2020-03-26T174254 117

6. After click on update qty, bundle item not render and total refund amount is 25$
Screenshot - 2020-03-26T174317 561

7. Create credit memo

Expected result (*)
Credit Memo for amount 25$ created

Actual result (*)

Error: "The most money available to refund is xx.xx.

Screenshot - 2020-03-26T174352 267
### Addition info:

After check issue, I saw this issue happen because the bundle item does not render in the form when all child product qty set to 0

Check function "canRefundItem" in Magento/Sales/Model/Order/CreditmemoFactory.php:145 When we create a refund for bundle product with all item qty = 0

`
if ($item->getHasChildren()) {
                foreach ($item->getChildrenItems() as $child) {
                    if (empty($qtys)) {
                        if ($this->canRefundNoDummyItem($child, $invoiceQtysRefundLimits)) {
                            return true;
                        }
                    } else {
                        if (isset($qtys[$child->getId()]) && $qtys[$child->getId()] > 0) {
                            return true;
                        }
                    }
                }
               return false;
            }
`

The return value will return false and exclude bundle parent item in return result

=> bundle parent item does not render in Magento/Sales/view/adminhtml/templates/order/creditmemo/create/items.phtml
So the qty value for each item not exist
Screenshot - 2020-03-26T174317 561

If the bundle item does not render in credit memo page and we create a credit memo, so when the function "createByOrder" to be call, the problem happen:
Magento/Sales/Model/Order/CreditmemoFactory.php:69

`        $qtyList = isset($data['qtys']) ? $data['qtys'] : [];`

 **$qtyList now is an empty array, but in this case, it should have data contain item_id with qty = 0**

 So this function will get all available refund qty for each item => the root of issue:
- If credit memo doesn't have adjustment amount, it will create credit memo for all item of bundle product (even we already updated all item qty to 0 - STEP 5 )

- If credit memo have adjustment amount, the error message will show: "The most money available to refund is xxxx"

I have pushed the commit to fix this issue
After apply update code, when we update qty for all bundle item to 0, the bundle item will show correctly:

Screenshot - 2020-03-26T182313 566

Result after create credit memo:

Screenshot - 2020-03-26T182334 431

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by KiuNguyen** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @KiuNguyen. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @KiuNguyen, here is your [Magento instance](https://i-27455-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-27455-2-4-develop.instances.magento-community.engineering/admin_ccb8](https://i-27455-2-4-develop.instances.magento-community.engineering/admin_ccb8)
Login: `3a5ccf76` Password: `d92ad8214142`
Instance will be terminated in up to 3 hours.

**Comment by KiuNguyen** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @KiuNguyen. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @KiuNguyen, here is your [new Magento instance](https://pr-27455.instances.magento-community.engineering).
Admin access: [https://pr-27455.instances.magento-community.engineering/admin_75f2](https://pr-27455.instances.magento-community.engineering/admin_75f2)
Login: `16c77c8c` Password: `20f715c3e072`
Instance will be terminated in up to 3 hours.

**Comment by nuzil** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @nuzil. Thank you for your request. I'm working on Magento instance for you

**Comment by nuzil** [comment]
@magento give me test instance

**Comment by magento-engcom-team** [comment]
Hi @nuzil. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @nuzil, here is your [new Magento instance](https://pr-27455.instances.magento-community.engineering).
Admin access: [https://pr-27455.instances.magento-community.engineering/admin_5563](https://pr-27455.instances.magento-community.engineering/admin_5563)
Login: `08f7c973` Password: `3e0a69104ac8`
Instance will be terminated in up to 3 hours.

File: app/code/Magento/Sales/Model/Order/CreditmemoFactory.php [hunk 1]
```diff
@@ -147,7 +147,7 @@ protected function canRefundItem($item, $qtys = [], $invoiceQtysRefundLimits = [
         if ($item->isDummy()) {
             if ($item->getHasChildren()) {
                 foreach ($item->getChildrenItems() as $child) {
-                    if (empty($qtys)) {
+                    if (empty($qtys) || (count(array_unique($qtys)) === 1 && end($qtys) == 0)) {
```
Comment (nuzil) [inline app/code/Magento/Sales/Model/Order/CreditmemoFactory.php]: Any particular reason why you do not want to make === compare?

Comment (KiuNguyen) [reply app/code/Magento/Sales/Model/Order/CreditmemoFactory.php]: Hi @nuzil 

> Any particular reason why you do not want to make === compare?

When I debug, I saw $qtys will return array value like this:

  array (size=3)
  48 => string '0' (length=1)
  49 => string '0' (length=1)
  50 => string '0' (length=1)

So now I already update code to use === compare

`if (empty($qtys) || (count(array_unique($qtys)) === 1 && (int)end($qtys) === 0)) {`

Please check this,
Thanks

**Review by nuzil — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @nuzil, thank you for the review.
[ENGCOM-7460](https://jira.corp.magento.com/browse/ENGCOM-7460) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
Before:
#27455Main
After:
Order information
#27455PR

image


Credit Memo information
image

**Review by nuzil — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @nuzil, thank you for the review.
[ENGCOM-7460](https://jira.corp.magento.com/browse/ENGCOM-7460) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
Result:
image

image

image

**Comment by vinhan2020** [comment]
give me 2.4-develop instance

**Pull request merged**